### PR TITLE
batch verification of BLS signature/msg pairs

### PR DESCRIPTION
**Pull request by tobowers**
This adds batch verification of a number of signatures where the messages are distinct. Generally this is just used for a performance enhancement. The benchmarks show that the verification is time for the 100 sigs here takes roughly the same amount as 50 signature verifications individually.

### CODE REVIEW

File: sign/bls/bls.go [hunk 1]
```diff
@@ -80,3 +114,15 @@ func hashToPoint(suite pairing.Suite, msg []byte) kyber.Point {
 	x := suite.G1().Scalar().SetBytes(h.Sum(nil))
 	return suite.G1().Point().Mul(x, nil)
 }
+
+func distinct(msgs [][]byte) bool {
+	m := make(map[[32]byte]bool)
+	for _, msg := range msgs {
+		h := sha256.Sum256(msg)
```
Comment (tobowers) [inline sign/bls/bls.go]: this takes about 1-2ms for the 100 sigs... I wonder if there's something faster here.

Comment (jeffallen) [reply sign/bls/bls.go]: a comment explaining why distinctness is required would be helpful here. And a test that confirms that repeated messages are correctly refused.

Comment (tobowers) [reply sign/bls/bls.go]: Good ideA

**Comment by tobowers** [comment]
No worries on the timing. Thanks! Happy holidays! :party:

**Comment by jeffallen** [comment]
@kc1212 , do you think that we should be implementing masking here, in the same way you did for BLS sigs in Java?

**Comment by kc1212** [comment]
I think the mask can be kept independent. In kyber, the mask is in the cosi package, which is actually not the right place because it's independent of cosi. We actually use cosi.NewMask as the BLS signature mask in skipchain. So the batch verification of bls can stay because it does some bls specific things like computing the pairing etc. If someone wants to use the mask to have fine-grained control, they'll use the mask method to aggregate the public keys and signatures (not possible for cosi).

File: sign/bls/bls.go (hunk 1 above)
Comment (tobowers) [reply sign/bls/bls.go]: I added a comment and a test

**Pull request merged**